### PR TITLE
update kmstool-enclave-cli with attested GenerateDataKey

### LINKED ISSUE #88: kmstool_enclave_cli need more feature
Such as GenerateDataKey so we can  ensure all data in security.

### LINKED ISSUE #92: kmstool_enclave_cli requires encryption algorithm when keyid is set
kms_tool_enclave_cli gained support for decrypt on ciphertext generated with asymmetric keys. This action requires both `keyID` and Encryption_Algorithm` to be set. The tool currently requires `Encryption_Algorithm` whenever a user sets `keyID` however, this is not required by the KMS API. When decrypting ciphertext generated with symmetric keys, for example, it is a best practice to explicitly specify the keyID (although it is not required) but this does not require setting the algorithm.

### PR DESCRIPTION
**Pull request by jdbean**
Based on work by Alex Richman

*Issue #, if available:* fixes #88 fixes #92

*Description of changes:* Adds, GenerateDataKey support to kmstool-enclave-cli and removes coupling of keyid and encryption_algorithm for decrypt operations.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by crockeea** [comment]
Looks good to me.

File: bin/kmstool-enclave-cli/main.c
Comment (crockeea) [inline bin/kmstool-enclave-cli/main.c:126]: Typo

Comment (jdbean) [reply bin/kmstool-enclave-cli/main.c]: typo fixed! Thank you @crockeea

File: bin/kmstool-enclave-cli/main.c
Comment (crockeea) [inline bin/kmstool-enclave-cli/main.c:64]: Did you intend to change this message? It's now missing the hint about being required when `key-id` is specified.

**Review by crockeea — DISMISSED** [review]

File: bin/kmstool-enclave-cli/main.c
Comment (jdbean) [reply bin/kmstool-enclave-cli/main.c:64]: Yes. This change was intentional as the requirement is not needed. See issue #92

**Review by crockeea — DISMISSED** [review]

File: bin/kmstool-enclave-cli/main.c
Comment (crockeea) [inline bin/kmstool-enclave-cli/main.c:106]: Could we add an option for `NumberOfBytes` to align with the [KMS API](https://docs.aws.amazon.com/kms/latest/APIReference/API_GenerateDataKey.html#KMS-GenerateDataKey-request-NumberOfBytes)? Users would have to specify `--key-spec` OR `num-bytes`.

File: bin/kmstool-enclave-cli/main.c
Comment (crockeea) [inline bin/kmstool-enclave-cli/main.c:374]: Is there a way to report the underlying error from KMS here? I'm getting this error (or the one on line 443), and I can't tell what the problem is.

Comment (jdbean) [reply bin/kmstool-enclave-cli/main.c:374]: At the moment I don't believe that the underlying C SDK is returning this data. While I agree that it would be a valuable enhancement, I don't think it should block merging this PR.

That said, it's possible that some additional information such as error codes are getting printed to console which would be visible in debug mode. I'd have to check. Additionally CloudTrail logs may provide some insight as well.

File: bin/kmstool-enclave-cli/main.c
Comment (jdbean) [reply bin/kmstool-enclave-cli/main.c:106]: At the moment I don't believe that the underlying C SDK has implemented this parameter. While I agree that it would be a valuable enhancement, I don't think it should block merging this PR.

**Review by alcioa — APPROVED** [review]

File: bin/kmstool-enclave-cli/main.c
Comment (alcioa) [reply bin/kmstool-enclave-cli/main.c:374]: Yeah, we plan at some point to pass KMS error logs to the caller since we would also like to improve the ACM for Nitro Enclaves user applications debugging in production.

**Comment by jdbean** [comment]
Thank you @alcioa and @crockeea for the reviews!

**Comment by sudhirj** [comment]
Thanks a ton for this, we're using Enclave system to generate data as well (data which can only be read again inside the enclave itself), and we've been scratching our heads over how to make sure that's locked down.

**Comment by jdbean** [comment]
Our pleasure @sudhirj !

**Comment by sudhirj** [comment]
Is there a doc somewhere on how to do this directly when forwarding requests to KMS? If we use the AWS Encryption SDK inside the Enclave, for example, assuming there is the vsock-proxy set up, will it automatically use attestation?

**Comment by jdbean** [comment]
@sudhirj 

> If we use the AWS Encryption SDK inside the Enclave, for example, assuming there is the vsock-proxy set up, will it automatically use attestation

This PR enables support for the kmstool-enclave-cli to make GenerateDataKey requests directly to KMS. At this time there is no direct integration between the AWS encryption SDK's AWS KMS Keyring features and the CLI tool ( and the underlying Nitro Enclaves SDK as well). I'd recommend opening an issue against the relevant AWS Encryption SDK repository requesting attestation and vsock enabled support for that tool.

**Review by crockeea — APPROVED** [review]

**Comment by jdbean** [comment]
@alcioa Do we need another reviewer on this? If so who should we tag? @petreeftime do you have the cycles to provide a review?

**Review by petreeftime — APPROVED** [review]

**Comment by sudhirj** [comment]
@jdbean Thanks a ton for landing this 👍🏾 Is it possible to cut a release with these changes? I'm going around updating scripts with a commit hash instead, would prefer to use version numbers.

**Comment by jdbean** [comment]
I'd recommend opening up an issue for this request with a reference to the PR.

**Pull request merged**
